charon/src/ast/types_utils.rs

```diff
@@ -518,3 +525,118 @@ impl<V> std::ops::DerefMut for VisitInsideTy<V> {
+        // This is the only possible instantation of [RegionBinder] inside a type so far
+        match item.downcast_mut::<PolyTraitDeclRef>() {
+            Some(_) => {
+                if enters {
+                    self.current_level += 1;
+                } else {
+                    self.current_level -= 1;
+                }
+                true
+            }
+            None => false,
+        }
+    }
+
+    fn try_subst_in_region(&mut self, item: &mut dyn std::any::Any) -> bool {
+        match item.downcast_mut::<Region>() {
+            Some(r) => {
+                match r {
+                    Region::BVar(db, id) => {
+                        if db.index == self.current_level {
+                            *r = self.region_subst.get(*id).unwrap().clone()
+                        }
+                    }
+                    _ => (),
+                };
+                true
+            }
+            None => false,
+        }
+    }
+
+    /// Returns [true] if the item is a binder (we use this to skip some checks)
+    fn try_subst_in_ty(&mut self, item: &mut dyn std::any::Any) -> bool {
+        match item.downcast_mut::<Ty>() {
+            Some(ty) => {
+                match ty.kind() {
+                    TyKind::TypeVar(id) => *ty = self.ty_subst.get(*id).unwrap().clone(),
+                    _ => (),
+                };
+                true
+            }
+            None => false,
+        }
+    }
+
+    fn try_subst_in_const_generic(&mut self, item: &mut dyn std::any::Any) -> bool {
+        match item.downcast_mut::<ConstGeneric>() {
+            Some(cg) => {
+                match cg {
+                    ConstGeneric::Var(id) => *cg = self.cg_subst.get(*id).unwrap().clone(),
+                    _ => (),
+                };
+                true
+            }
+            None => false,
+        }
+    }
+
+    fn try_subst_in_trait_ref(&mut self, item: &mut dyn std::any::Any) -> bool {
+        match item.downcast_mut::<TraitRef>() {
+            Some(tr) => {
+                match &mut tr.kind {
+                    TraitRefKind::Clause(id) => *tr = self.trait_subst.get(*id).unwrap().clone(),
+                    _ => (),
+                };
+                true
+            }
+            None => false,
+        }
+    }
+}
+
+impl VisitorMut for Subst {
+    fn visit(&mut self, item: &mut dyn std::any::Any, event: Event) {
+        // Check if this is a binder
+        let enters = matches!(event, Event::Enter);
+        if self.try_enter_or_exit_binder(item, enters) {
+        }
+        // Other cases: we only do something upon entering
+        else if enters {
+            if self.try_subst_in_region(item) {
+            } else if self.try_subst_in_ty(item) {
+            } else if self.try_subst_in_const_generic(item) {
+            } else {
+                let _ = self.try_subst_in_trait_ref(item);
+            }
+        }
+    }
```

Comment:
This won't recurse deeply into `Ty` because the `DriveMut` impl of `Ty` does nothing by default. That's a good thing, otherwise a visitor like this one would not be possible. To make it work you need to call `ty.drive_inner_mut()` whenever you want to recurse. But then you don't need to downcast by hand, use `#[derive(VisitorMut)]` instead, you can see examples in many of the micro-passes.

To avoid double-substitutions, simply substitute on the `Exit` event instead of the `Enter` one.

Reply:
Ah good `#[derive(VisitorMut)]` looks like what I'm looking for.
So what's the mental model for `Enter` and `Exit`? Does the visitor do: "call the function on enter", "drive inside the type", "call the function on exit"?

Reply:
yep, exactly. also instead of calling `drive_inner_mut`, you can wrap your visitor in `Ty::visit_inside_ty` which will do it for you.

Reply:
(I updated the code)